### ISSUE TITLE
Evaluate global constructors (life before main)

### ISSUE THREAD

**Issue by oli-obk** [solson/miri#450]
@eddyb is lawful evil https://github.com/neon-bindings/neon/blame/master/src/lib.rs#L49

basically we should look in platform specific link sections for statics and run their value before anything else

```rust
#[cfg_attr(target_os = "linux", link_section = ".ctors")]
#[cfg_attr(target_os = "macos", link_section = "__DATA,__mod_init_func")]
#[cfg_attr(target_os = "windows", link_section = ".CRT$XCU")]
pub static __FOOMP: extern "C" fn() = {
    extern "C" fn __foomp() {
        // life before main goes here
    }
    __foomp
}
```

**Comment by Ericson2314** [comment]
🤮 @eddyb!

**Comment by RalfJung** [comment]
Are you sure about the "lawful" part? :P

More seriously though -- shouldn't we instead lobby for some way to express this in Rust proper? I already feel bad about that TLS emulation we are doing...

**Comment by eddyb** [comment]
> shouldn't we instead lobby for some way to express this in Rust proper?

No, Rust as a language is uninterested in life-before-main. The only reason any of this works is because it's an OS feature we can't really stop you from opting into.

**Comment by RalfJung** [comment]
See https://github.com/rust-lang/rust/issues/66862#issuecomment-559862805 for more information on these sections.

**Comment by joshtriplett** [comment]
You might find the following C program useful for experimenting with init section order:

```c
#include <stdio.h>

typedef void (*FN)(void);

#define MKINIT(function, section) \
    void function(void) { puts(section); } \
    __attribute__((__section__(section))) FN function##_section[] = { function };

MKINIT(init_array, ".init_array")
MKINIT(init_array_00001, ".init_array.00001")
MKINIT(init_array_00002, ".init_array.00002")
MKINIT(init_array_65533, ".init_array.65533")
MKINIT(init_array_65534, ".init_array.65534")
MKINIT(fini_array, ".fini_array")
MKINIT(fini_array_00001, ".fini_array.00001")
MKINIT(fini_array_00002, ".fini_array.00002")
MKINIT(fini_array_65533, ".fini_array.65533")
MKINIT(fini_array_65534, ".fini_array.65534")
MKINIT(ctors, ".ctors")
MKINIT(ctors_00001, ".ctors.00001")
MKINIT(ctors_00002, ".ctors.00002")
MKINIT(ctors_65533, ".ctors.65533")
MKINIT(ctors_65534, ".ctors.65534")
MKINIT(dtors, ".dtors")
MKINIT(dtors_00001, ".dtors.00001")
MKINIT(dtors_00002, ".dtors.00002")
MKINIT(dtors_65533, ".dtors.65533")
MKINIT(dtors_65534, ".dtors.65534")

int main(void)
{
    puts("main");
    return 0;
}
```

This prints:
```
.ctors.65534
.init_array.00001
.ctors.65533
.init_array.00002
.ctors.00002
.init_array.65533
.ctors.00001
.init_array.65534
.init_array
.ctors
main
.dtors
.fini_array
.fini_array.65534
.dtors.00001
.fini_array.65533
.dtors.00002
.fini_array.00002
.dtors.65533
.fini_array.00001
.dtors.65534
```

Note that *weird* things happen if you try using `00000` or `65535` or numbers outside that range, or non-numeric suffixes; the entire list gets sorted strangely if you do that. I don't think you need to worry about reproducing that exact behavior, though; I think that's a linker script quirk that doesn't normally come up because GCC never emits sections with such "weird" suffixes. Also note that the aforementioned weirdness may differ between GNU ld and LLVM's lld.

**Comment by RalfJung** [comment]
Thanks! We'll need to turn this into a Rust program as a test for this feature, but that shouldn't be too hard I guess.

> Note that weird things happen if you try using 00000 or 65535 or numbers outside that range, or non-numeric suffixes; the entire list gets sorted strangely if you do that. I don't think you need to worry about reproducing that exact behavior, though; I think that's a linker script quirk that doesn't normally come up because GCC never emits sections with such "weird" suffixes. Also note that the aforementioned weirdness may differ between GNU ld and LLVM's lld.

Sounds like it would be best to just make Miri error in those cases.

**Comment by oli-obk** [comment]
We also need to bail if two entries for the same link section exist

**Comment by joshtriplett** [comment]
@oli-obk No, you can have arbitrarily many entries in a section; it's an array of function pointers. Entries for the same section get concatenated in link order.

**Comment by RalfJung** [comment]
How does one detect the length of the array? Do link sections have a notion of "length"?

**Comment by oli-obk** [comment]
> Entries for the same section get concatenated in link order.

I don't think miri can figure out link order. For now we can bail out on multiple entries for the same section, but in the future we can check that they are order independent instead

**Comment by RalfJung** [comment]
>  in the future we can check that they are order independent instead

That sounds like fantasy to me. ;) I don't see a feasible implementation strategy for this.

**Comment by eddyb** [comment]
> Do link sections have a notion of "length"?

@RalfJung yes, in some sense. AFAIK, linker scripts can define a symbol before and one after, so that code could iterate between those two symbols.

When you're operating before linking (as miri would need to do), any `static` in a relevant section is effectively a part of the final array, so it must contain 0, 1 or more `fn` pointers.

That is, ignoring the Rust type of the `static`, take the resulting memory allocation and divide it into pointer-sized elements that you then read out as `fn` pointers.

I don't think link order is a serious concern, and other than the numeric priorities which @joshtriplett explained, you can just pick some visiting order (and maybe lint the fact that a linker might differ?).

**Comment by RalfJung** [comment]
@oli-obk @bjorn3 do you know a good way to get "all statics that are in a particular linker section"?

I suspect this issue may be responsible for https://github.com/rust-lang/rust/issues/123583. Maybe it's time to stop using this ad-hoc hack for our Windows thread-local destructors.

**Comment by RalfJung** [comment]
Oh, neat!

> it requires a hack in the exported_symbols query override to also preserve the Instances of all #[used] symbols.

Does the regular query do that and Miri just lost it? Or do we somehow have to do something the regular query does not do? I assume regular rustc also has to preserve these statics for codegen as well...

**Comment by bjorn3** [comment]
> Does the regular query do that and Miri just lost it?

The regular query doesn't do that.

> I assume regular rustc also has to preserve these statics for codegen as well...

Statics are codegened in the crate that defines them and as such don't need to end up in `exported_symbols` if they aren't exported from the current crate (and not used by an exported inlinable function). For calling global constructors however miri would need all statics even if they aren't exported for as long as they are marked as `#[used]`.

**Comment by ChrisDenton** [comment]
I think it'd be better for the compiler to have some basic support for platform specific linker magic. The library code is already a bit too hacky for my liking so it'd be an improvement there too.

**Comment by RalfJung** [comment]
Yeah it would be better but also a lot more work to implement. I won't stop anyone who tries to do that but I don't have the resources to do it.

> Statics are codegened in the crate that defines them and as such don't need to end up in exported_symbols if they aren't exported from the current crate (and not used by an exported inlinable function). For calling global constructors however miri would need all statics even if they aren't exported for as long as they are marked as #[used].

The static in question is not marked `used`

```rust
#[link_section = ".CRT$XLB"]
pub static p_thread_callback: unsafe extern "system" fn(c::LPVOID, c::DWORD, c::LPVOID) =
    on_tls_callback;
```
So there's something here I don't understand.

> Statics are codegened in the crate that defines them

The query we override runs in that crate, so why is that relevant?

For finding extern functions with a link_name we also just do exactly what codegen does, I think. So it seems strange to me that here we'd have to do more than what codegen does.

**Comment by joboet** [comment]
> So there's something here I don't understand.

See rust-lang/rust#121596 for why the `#[used]` was removed.

**Comment by bjorn3** [comment]
> > Statics are codegened in the crate that defines them
> 
> The query we override runs in that crate, so why is that relevant?

We need to override the query to include `#[used]` items because otherwise they wouldn't be encoded in the crate metadata as exported symbols and thus have no way to find all used statics for dependencies.

> For finding extern functions with a link_name we also just do exactly what codegen does, I think. So it seems strange to me that here we'd have to do more than what codegen does.

When doing codegen the static initializer ends up an object file and codegen doesn't have to care about it anymore. Only the linker needs to find it. Miri however will need to know all static initializers of all dependencies without inspecting any (non existent) object files. As such the list of static initializers needs to be encoded in the crate metadata, which I did by ensuring they end up in the exported_symbols query.

**Comment by RalfJung** [comment]
> When doing codegen the static initializer ends up an object file and codegen doesn't have to care about it anymore. Only the linker needs to find it. Miri however will need to know all static initializers of all dependencies without inspecting any (non existent) object files. As such the list of static initializers needs to be encoded in the crate metadata, which I did by ensuring they end up in the exported_symbols query.

I see... well I think I see. ;)

But it seems to me Miri should preserve all statics with a link_name, not all statics with `#[used]`. That is what we do for functions as well, isn't it? (I don't see a check for `link_name` in our `exported_symbols` query, but we only really care about the ones with a `link_name`.)

**Comment by bjorn3** [comment]
We do preserve those with link_name already, but not those with link_section.

> I don't see a check for link_name in our exported_symbols query, but we only really care about the ones with a link_name.

I believe that is covered by the `contains_extern_indicator` check.

**Comment by RalfJung** [comment]
Ah right I mixed up link_name and link_section.

Sounds like a non-used static with a magic link_section has a really strange semantics - the function may or may not be called based on optimizations. That's very fragile and tricky to emulate in Miri. I would prefer if we didn't do hacks like that...

**Comment by ChrisDenton** [comment]
Which is one reason why I'd prefer proper compiler support for TLS destructors 😉

But it is just an optimization. Considering it to be always called is fine.

**Comment by RalfJung** [comment]
> But it is just an optimization. Considering it to be always called is fine.

What Miri should be doing is randomly sometimes call it and sometimes not. That reflects the fact that it is unspecified whether the static ends up in the final binary or not.

> Which is one reason why I'd prefer proper compiler support for TLS destructors 😉

I don't disagree with that. But I strongly disagree with using hacks relying on unspecified behavior to work around language limitations. The fix for a missing language feature is to add the feature. If something can't be done properly then doing it improperly is not a viable alternative IMO.

**Comment by RalfJung** [comment]
In this case I think the worst thing that can happen is memory leaks. Or can unsafe code somehow rely on TLS destructors being called? If yes then I would say https://github.com/rust-lang/rust/pull/121596 introduced a soundness issue, as we no longer can guarantee that TLS destructors are called.

**Comment by ChrisDenton** [comment]
I'm not sure I understand your point. If there is ever a TLS dtor then it will be called. Always.

If there is not then it's a no-op in any case so is eligible for removal.

**Comment by RalfJung** [comment]
> I'm not sure I understand your point. If there is ever a TLS dtor then it will be called. Always.

You are relying on the completely unspecified assumption that these volatile reads "pull in" the static.

**Comment by ChrisDenton** [comment]
Sure but it's not the only place in the standard library that does that. In fact within that very function there's a volatile read being used to pull in another symbol `_tls_used`).

**Comment by RalfJung** [comment]
Relying on unspecified behavior more than once doesn't make it any or correct to do so.

**Comment by ChrisDenton** [comment]
`#[used]` does not provide the semantics we want either. It only guarantees that a symbol will end up in an object file. It does not guarantee that the linker includes that object in the final binary. This can be a real concern, depending on how things end up being split. I've encountered this issue when trying to add a `.drectve` section in a library.

**Comment by RalfJung** [comment]
That sounds like a bug in `#[used]`, maybe caused by limitations of linkers? If linkers have a way to force an object to persist, `#[used]` should set that.

But fixing `#[used]` is orthogonal to the issue in this thread.

**Comment by ChrisDenton** [comment]
It's documented behaviour. https://doc.rust-lang.org/reference/abi.html#the-used-attribute

> This attribute forces the compiler to keep the variable in the output object file

But my point is that without compiler support there are no right options for including the global destructor.

**Comment by petrochenkov** [comment]
What exactly it is supposed to be used for, by the way?
I didn't read the whole thread, but in general ctor/dtor sections are supposed to be marked as GC roots by linker automatically, and not removed.

**Comment by bjorn3** [comment]
> What happened with #[used(linker)]?

It hasn't been stabilized yet. Note that the LLVM docs recommend only exposing what we expose as `#[used(linker)]` to end users and not `#[used(compiler)]` (which is what `#[used]` is currently implemented as)

**Comment by bjorn3** [comment]
> However, object files aren't guaranteed to be inspected by the linker if they're in a lib and don't contain a needed symbol.

IMO we should link all rlibs as `--whole-archive`, which would avoid this problem.

**Comment by RalfJung** [comment]
> How is that unspecified? read_volatile is guaranteed not to be optimized away, so there will always be a reference to the relevant symbol. And all referenced symbols must be placed in the final binary, right?

read_volatile guarantees to actually perform a hardware read from the given address. It says nothing about symbols appearing in link sections. What you are using here is some sort of accident, not a result of any kind of guarantee or design.

**Comment by joboet** [comment]
> read_volatile guarantees to actually perform a hardware read from the given address.

But that alone guarantees that the symbol must be present, right? Otherwise you couldn't get that address.

Is there a sound (hypothetical) optimization in either compiler or linker that could remove the symbol if it is being referenced from other used things?

**Comment by RalfJung** [comment]
We don't have a definition of how "used" symbols are defined (except by adding the attribute). So there's no starting point from which you can deduce anything, with or without volatile. Optimizations are irrelevant.

Try putting what you are asking for into an opsem. Basically when Miri starts it needs to *predict* whether a volatile access will happen in this execution, and only then consider the symbol to be in the link section. That is obviously unworkable, and yet it is what the standard library currently expects. And even that is changing the spec by explicitly saying that volatile matters for symbol existence!

We can also just always call the function but that is clearly not reflecting reality as it makes the volatile make no difference.

**Comment by joboet** [comment]
> We don't have a definition of how "used" symbols are defined (except by adding the attribute). So there's no starting point from which you can deduce anything, with or without volatile.

That seems like it should be specified.

>  Optimizations are irrelevant.

This whole debate is just about optimizations. `std` works fine with or without the optimization of removing an unused symbol.

> Try putting what you are asking for into an opsem. Basically when Miri starts it needs to *predict* whether a volatile access will happen in this execution, and only then consider the symbol to be in the link section. 

It doesn't need to predict anything, because IIUC miri is not about proving any optimization sound, but about proving that the code conforms to the rules needed to justify those optimizations. Just like how miri doesn't need to predict if two function pointers will end up equal or whether two stack allocations will end up at the same address, it doesn't need to predict whether the symbol will be removed.

> That is obviously unworkable, and yet it is what the standard library currently expects.

No, `std` just expects that the symbol will be linked if it is referenced.

> And even that is changing the spec by explicitly saying that volatile matters for symbol existence!

But it does matter! A symbol referenced by code that is used must itself be used. If the `read_volatile` call exists (which, if it is reachable, it always is), then the symbol must exist, otherwise the `read_volatile` wouldn't be valid (what pointer should it read from otherwise?).

> We can also just always call the function but that is clearly not reflecting reality as it makes the volatile make no difference.

That is totally valid, though! `std` doesn't rely on LLVM removing the symbol if it is unused.

**Comment by RalfJung** [comment]
> This whole debate is just about optimizations. std works fine with or without the optimization of removing an unused symbol.

No, the whole debate is about what we guarantee for Rust program semantics. The spec does not mention optimizations. That's the entire point of the "behave as-if it was executed on the Abstract Machine" approach.

> It doesn't need to predict anything, because IIUC miri is not about proving any optimization sound, but about proving that the code conforms to the rules needed to justify those optimizations. Just like how miri doesn't need to predict if two function pointers will end up equal or whether two allocations will end up at the same address, it doesn't need to predict whether the symbol will be removed.

Whenever the symbol is *allowed* to be removed, Miri should make a non-deterministic choice to remove it. That's wht it means for Miri to precisely implement the spec. The choice needs to be made when the program starts so that the answer to whether the symbol exists is consistent.

You definition of "allowed to be removed" seems to involve whether during the execution of the program there will be a volatile access to the symbol. The only way to decide that at program startup is to predict the future. So you are proposing a spec that requires clairvoyance. (This is angelic choice, i.e. it is not without precedent in the realm of the spec. But it makes tools like Miri impossible which is why IMO we should avoid doing such things in the spec whenever we can.)

> That is totally valid! std doesn't rely on LLVM removing the symbol if it is unused.

It is *valid*, sure. But that's not what we are discussing here. What we are discussing here is ensuring Miri can cover all allowed executions of the program, so that no matter what optimizations do, there's some chance Miri will behave the way the program will. That's why Miri picks a "random" schedule and "random" allocation base addresses and gives compare_exchange_weak a "random" chance to spuriously fail. It is simply not enough to find *a* valid execution; specifying Rust means exhaustively defining *all* valid executions.

Clearly for some cases it is valid for the symbol to not exist. So far the only spec you have indicated that explains when *exactly* the symbol *must* exist involves predicting the future. I'm not happy with that spec.

**Comment by joboet** [comment]
I'll split my answer into three parts:

### miri's current behaviour is unsound

miri currently performs a load of the `static` and thinks that the `static` will be removed. Those two things cannot be valid at the same time. After all, if the `static` were removed, what address would the load use?

### When is the symbol allowed to be removed

The final rule is quite simple: A symbol is allowed to be removed unless it is reachable, where
1. Every symbol that is marked as `#[used]` or implicitly reachable (like `main`) is reachable.
2. Every symbol referenced by a reachable symbol is itself reachable.

The problem lies in what that "referenced" means, because a non-volatile load for instance is insignificant and can be removed if its result is unused. In our example the `read_volatile` must not be removed (because it is volatile), so if it is itself reachable, then so must the `static` be.

### Can miri cover all possible executions

I leave this up to you but currently it doesn't. The following code is unsound, not because of the first `unreachable_unchecked` but because of the second. I know this because there is no possible optimization that could result in `a`/`b` and `c` being the same function, but `a` and `b` can (and will) be merged. miri will not catch the UB as doing so would involve modelling every optimization allowed under the spec to determine whether the functions will be merged.

```rust
#[inline(never)]
extern "C" fn a(x: i32, y: i32) -> i32 {
    x + y
}

#[inline(never)]
extern "C" fn b(x: i32, y: i32) -> i32 {
    x + y
}

#[inline(never)]
extern "C" fn c(x: i32, y: i32) -> i32 {
    x - y
}

fn main() {
    println!("a: {}", a as usize);
    println!("b: {}", b as usize);
    println!("c: {}", c as usize);

    if a as usize == c as usize {
        unsafe { std::hint::unreachable_unchecked() } // Can never be reached, this is sound.
    }

    if a as usize == b as usize {
        unsafe { std::hint::unreachable_unchecked() } // Can be reached, this is unsound.
    }
}
```

**Comment by bjorn3** [comment]
> there is no possible optimization that could result in a/b and c being the same function

Neither function is actually called, so there is no reason optimizations can't replace all arguments with hard coded 0, which results in the return value to be a fixed 0 for a, b and c. LLVM can and does remove arguments if all callers (in this case all 0 callers) are guaranteed to pass the same value.

**Comment by RalfJung** [comment]
> miri currently performs a load of the static and thinks that the static will be removed. Those two things cannot be valid at the same time. After all, if the static were removed, what address would the load use?

Miri's current behavior is a gross hack, there's nothing useful to be gleamed from it.

> When is the symbol allowed to be removed

The actual question is when is the symbol required to exist.

> The final rule is quite simple

It's not simple. You are doing the equivalent of "why don' you *just*". Please be open to the idea that when t-opsem folks tell you the opsem is complicated, then it actually is.

> Every symbol referenced by a reachable symbol is itself reachable.

"reachable" cannot be relevant to the op.sem as it is an optimization-dependent notion. I don't see any way that "reachable" can be useful for defining the guarantees of program behavior.

**Comment by joboet** [comment]
> And then it diverted into discussing whether what `std` does is actually correct under any plausible opsem that t-opsem would be happy with.

I realize this discussion has nothing to do with the issue title anymore, sorry about that. Does GitHub have a move feature?

> > When is the symbol allowed to be removed
> 
> I would phrase it as "when is the symbol required to exist", but in the send that's equivalent.

Absolutely, let's use your phrasing.

> > The final rule is quite simple
> 
> It's not simple. You are doing the equivalent of "why don' you _just_". Please be open to the idea that when t-opsem folks tell you the opsem is complicated, then it actually is.
> 
> > Every symbol referenced by a reachable symbol is itself reachable.
> 
> "reachable" cannot be relevant to the op.sem as it is an optimization-dependent notion. I don't see any way that "reachable" can be useful for defining the guarantees of program behavior. The same goes for "referenced".

I'm really sorry for wasting your time and typing, I was thinking about this all wrong. I hope I got it know.

The thing is that I'd really like for what `std` does to be valid. The `read_volatile` on the destructor function is just a neat optimization for binary size, but the `read_volatile` on `_tls_used` implying it's presence is actually required for the code to work.

So in light of that, I'd advocate for miri to amend its usedness rules. My first instinct was to consider every `static` that is read from used, but that makes loads side-effecting, which I highly doubt you'd want to have.

But this rule I think works: *A volatile load or store to a `static` makes that `static` used.*

Volatile operations are already side-effecting so this shouldn't impact anything else. `std`s case is fixed because we always do the `read_volatile` before destructors are run. It doesn't work for similar hacks for global constructors, but I guess that is really unavoidable (and those hacks aren't really guaranteed to work either if the dynamic linker does some weird dynamic loading of symbols). And the LLVM lowering already respects this rule, because LLVM will not remove statics when there is a `read_volatile` of them that could be called.

**Comment by RalfJung** [comment]
> But my point is that without compiler support there are no right options for including the global destructor.

There are options that are different degrees of wrong, and IMO https://github.com/rust-lang/rust/pull/121596 made it more wrong than before. Now you're not only relying on whatever happens between the object file and the final binary keeping the symbol around, you are also relying on this `volatile` trick that has no proper justification.

IOW, there are two stages where the symbol can get lost, and your argument seems to be "because the second stage is lossy, it's okay to not use the attribute that makes the first stage definitely not-lossy". That's not a valid argument; we have been discussing all along what happens in the first stage (whether the symbol makes it to the object file) so the second stage (whether the symbol makes it to the binary) is entirely irrelevant.

**Comment by petrochenkov** [comment]
>IMO https://github.com/rust-lang/rust/pull/121596 made it more wrong than before

FWIW, I agree with this statement.
I've seen that PR, but didn't have time to investigate and interfere.
If `#[used]`/`#[used(compiler)]`/`#[used(linker)]` don't work, then they need to be fixed to work.
We have ways to both keep object files alive (symbols.o) and keep sections non-garbage-collectable (RETAIN sections), if they don't work on some platforms then other ways to make it work are likely possible.

**Comment by RalfJung** [comment]
With https://github.com/rust-lang/rust/pull/123937 now has the basic infrastructure to deal with link_sections that are actually arrays, which should make it a lot easier to implement support for constructors as well.

**Comment by RalfJung** [comment]
Mostly implemented by https://github.com/rust-lang/miri/pull/4459 :)
